src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.DecisionDagRewriter.cs

```diff
@@ -865,59 +870,237 @@ private void EnsureStringHashFunction(int labelsCount, SyntaxNode syntaxNode)
                 privateImplClass.TryAddSynthesizedMethod(method.GetCciAdapter());
             }
 
-            private void LowerWhenClause(BoundWhenDecisionDagNode whenClause)
+#nullable enable
+            private void LowerWhenClauses(ImmutableArray<BoundDecisionDagNode> sortedNodes)
             {
-                // This node is used even when there is no when clause, to record bindings. In the case that there
-                // is no when clause, whenClause.WhenExpression and whenClause.WhenFalse are null, and the syntax for this
-                // node is the case clause.
-
-                // We need to assign the pattern variables in the code where they are in scope, so we produce a branch
-                // to the section where they are in scope and evaluate the when clause there.
-                var whenTrue = (BoundLeafDecisionDagNode)whenClause.WhenTrue;
-                LabelSymbol labelToSectionScope = GetDagNodeLabel(whenClause);
-
-                ArrayBuilder<BoundStatement> sectionBuilder = BuilderForSection(whenClause.Syntax);
-                sectionBuilder.Add(_factory.Label(labelToSectionScope));
-                foreach (BoundPatternBinding binding in whenClause.Bindings)
-                {
-                    BoundExpression left = _localRewriter.VisitExpression(binding.VariableAccess);
-                    // Since a switch does not add variables to the enclosing scope, the pattern variables
-                    // are locals even in a script and rewriting them should have no effect.
-                    Debug.Assert(left.Kind == BoundKind.Local && left == binding.VariableAccess);
-                    BoundExpression right = _tempAllocator.GetTemp(binding.TempContainingValue);
-                    if (left != right)
+                // The way the DAG is prepared, it is possible for different `BoundWhenDecisionDagNode` nodes to
+                // share the same `WhenExpression` (same `BoundExpression` instance).
+                // So we can't just lower each `BoundWhenDecisionDagNode` separately, as that would result in duplicate blocks
+                // for the same `WhenExpression` and such expressions might contains labels which must be emitted once only.
```

Comment:
Out of curiosity, would it make sense to rewrite when-expressions and replace offending nodes? similar to spilling but for jumps and labels. Just wondering if a nested dispatch could be avoided.

Reply:
Yes, that's definitely an option. 

Three broad approaches were considered:
1. avoid the bifurcation with duplication in the first place (we need branches of the DAG to join back together, this likely involves replacing the selected pattern test with checking the cached result of previous test)
2. using subroutine/dispatch approach to share body of when-expressions (or extracting share-when expressions to a local function).
3. properly cloning such shared BoundExpressions rather than re-using instances

In terms of trade-offs, options 1 is the most difficult to land safely and successfully for 17.0 timeframe.
As for choosing option 2 over 3, one reason is when-expressions can be arbitrarily complex so duplicating them could inflate the IL. Also we'd have to review all bound nodes (are BoundLabelStatements the only nodes that need adjustment when cloning?).

Reply:
>(we need branches of the DAG to join back together, this likely involves replacing the selected pattern test with checking the cached result of previous test)

If I understand that correctly, you mean to capture the when-expression in a temp? That wouldn't help because when-clauses are always located right before a leaf node. We *could* merge them by reordering states (i.e. running a when-clause before its preceding pattern) but that will obviously break the order of evaluation, so I think the only remaining option which avoids duplication and preserve the existing evaluation order is the one currently taken.